### PR TITLE
Reworked API

### PR DESCRIPTION
**Pull request by shalupov**

### CODE REVIEW

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt [hunk 1]
```diff
@@ -123,20 +231,52 @@ interface BuildConfiguration {
...
-    fun fetchBuildTags(): List<String>
+    val buildTags: List<String>
+    val finishBuildTriggers: List<FinishBuildTrigger>
+    val artifactDependencies: List<ArtifactDependency>
 
-    fun fetchFinishBuildTriggers(): List<FinishBuildTrigger>
+    fun setParameter(name: String, value: String)
 
+    fun runBuild(parameters: Map<String, String>? = null,
+                 queueAtTop: Boolean? = null,
```
Comment (chashnikov) [inline teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: What does `null` means for these boolean parameters?

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt [hunk 2]
```diff
@@ -263,6 +448,27 @@ interface Revision {
...
+    UNKNOWN,
+}
+
+interface TestOccurrence {
+    val name : String
+    val status: TestStatus
+    val duration: Long
+    val details : String
+    val ignored: Boolean
+    val currentlyMuted: Boolean
```
Comment (chashnikov) [inline teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: What's difference between `muted` and `currentlyMuted`?

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt [hunk 3]
```diff
@@ -263,6 +448,27 @@ interface Revision {
...
+    IGNORED,
+    FAILED,
+
+    UNKNOWN,
+}
+
+interface TestOccurrence {
+    val name : String
+    val status: TestStatus
+    val duration: Long
```
Comment (chashnikov) [inline teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: `duration` is in milliseconds, right? It makes sense to mention this in the name or a KDoc.

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt [hunk 4]
```diff
@@ -77,41 +131,95 @@ interface BuildLocator {
...
+    fun getHomeUrl(branch: String? = null): String
+    fun getTestHomeUrl(testId: TestId): String
+
+    val childProjects: List<Project>
+    val buildConfigurations: List<BuildConfiguration>
+    val parameters: List<Parameter>
+
+    fun setParameter(name: String, value: String)
+
+    fun createVcsRoot(id: VcsRootId, name: String, type: VcsRootType, properties: Map<String, String>): VcsRoot
```
Comment (chashnikov) [inline teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: This violates [item 2 from the contribution guideline](https://github.com/JetBrains/teamcity-rest-client/blob/master/CONTRIBUTING.md). What should I pass here if I want to create a Git VCS root with a specific fetch URL?
I suggest to add a method `createGitVcsRoot` and pass the main parameters (fetch URL, default branch, branch spec, maybe authentication settings?) as explicit named arguments to it. You can keep `additionalProperties` argument for other properties, and also you can leave this method to support creation of VCS roots for other VCSes.

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt [hunk 5]
```diff
@@ -145,55 +285,79 @@ interface Branch {
...
+}
+
 interface Build {
     val id: BuildId
-    val buildTypeId: BuildConfigurationId
-    val buildNumber: String
-    val status: BuildStatus
+    val buildConfigurationId: BuildConfigurationId
+    val buildNumber: String?
+    val status: BuildStatus?
```
Comment (chashnikov) [inline teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: `status` and `state` look similar to me. Maybe we can merge them into a single enum?

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt [hunk 6]
```diff
@@ -145,55 +285,79 @@ interface Branch {
...
+    val statusText: String?
+    val queuedDate: Date
+    val startDate: Date?
+    val finishDate: Date?
 
-    fun fetchParameters(): List<Parameter>
+    val parameters: List<Parameter>
 
-    fun fetchRevisions(): List<Revision>
+    val revisions: List<Revision>
```
Comment (chashnikov) [inline teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: What does `revisions` mean? The latest revisions from VCS roots attached to this build configuration? It makes sense to add KDoc here and maybe rename the property.

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt [hunk 7]
```diff
@@ -77,41 +131,95 @@ interface BuildLocator {
...
+
+    val childProjects: List<Project>
+    val buildConfigurations: List<BuildConfiguration>
+    val parameters: List<Parameter>
+
+    fun setParameter(name: String, value: String)
+
+    fun createVcsRoot(id: VcsRootId, name: String, type: VcsRootType, properties: Map<String, String>): VcsRoot
+    fun createProject(id: ProjectId, name: String): Project
+    fun createBuildConfiguration(buildConfigurationDescriptionXml: String): BuildConfiguration
```
Comment (chashnikov) [inline teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: We need to give some information about format of `buildConfigurationDescriptionXml`, at least link to some documentation. Currently it's absolutely obscure.

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt (hunk 5 above)
Comment (shalupov) [reply teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: Nope. `state` tell us current point of a build lifecycle . `status` tell us a build result.
Also at some point of time we'll need to support failed to start builds

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt (hunk 7 above)
Comment (shalupov) [reply teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: I'll add a javadoc how to get it, but not much more.

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt (hunk 1 above)
Comment (shalupov) [reply teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: It means default value on TeamCity. But in this case it's probably better to be specific

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt (hunk 4 above)
Comment (shalupov) [reply teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: Yeah, and authentication settings, 4 cases of them at least. There isn't a simple configuration of a VCS root. I've added a javadoc.

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt (hunk 6 above)
Comment (shalupov) [reply teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: Added kdocs for revisions and changes

File: teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt (hunk 5 above)
Comment (chashnikov) [reply teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: But `status` makes sense only if `state` is `FINISHED`, isn't it? Or `status` may be `ERROR` while a build is still running?

Comment (shalupov) [reply teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt]: yes, the status is present after a build was started

**Pull request merged**